### PR TITLE
[Merged by Bors] - Use NodePorts for local environment

### LINKED ISSUE #668: Use `nodePort` rather than `tunnel` for accessing services in the minikube
We use [minikube tunnel](https://minikube.sigs.k8s.io/docs/handbook/accessing/) for exposing SC/SPU services to outside cluster.  It however has unfortunately downside that it's difficult to setup and maintain due to different ways to set up depends on OS.  

Alternative is to use `nodeport` which minikube supports.   It also has benefit of being more portable to other K8 setup such as [kind](https://kind.sigs.k8s.io) or [k3](https://rancher.com/products/k3s/).  

This issue is about experimentation to see if it is possible to migrate to `nodeport`

### PR DESCRIPTION
**Pull request by tjtelan**
(Continuing from #1047)
Closes #1047 

Resolves #668

NodePort is now the default service type for `fluvio cluster start`.

SC startup:
* Check the k8s `service` type. If it is NodePort,
  * If `service` is NodePort, lookup to the k8s `node` and use the first node's IP to construct the public address for SC.
  * If `service` is NodePort, use the `service` NodePort value instead of the container port.
  * When starting the SPUs with `helm`, set `fluvio.io/ingress-address` service annotation with the public address for SPUs to use.

CI tests only use `minikube tunnel` in the cluster upgrade test, but it can be removed in the release following `0.8.4`. - [tracking issue](https://github.com/infinyon/fluvio/issues/1141)

### CODE REVIEW

File: k8-util/helm/fluvio-app/values.yaml [hunk 1]
```diff
@@ -6,7 +6,7 @@ cloud: minikube
 loadBalancer:
   serviceAnnotations: {}
 service:
-  type: LoadBalancer
+  type: NodePort
```
Comment (nicholastmosher) [inline k8-util/helm/fluvio-app/values.yaml:9]: Should the helm chart be optimized for the local use-case or for the production use-case? In one of those cases the helm invocation will need to overwrite this.

File: src/cluster/src/cli/start/k8.rs [hunk 2]
```diff
@@ -34,6 +34,7 @@ pub async fn process_k8(
         .chart_values(opt.k8_config.chart_values)
         .render_checks(true)
         .upgrade(upgrade)
+        .use_cluster_ip(true)
```
Comment (nicholastmosher) [inline src/cluster/src/cli/start/k8.rs:37]: If your branch is based on mine then you should double-check if this is needed. I added this when I was still figuring out the different IP types.

File: src/cluster/src/start/k8.rs [hunk 3]
```diff
@@ -355,6 +361,14 @@ pub struct ClusterConfig {
     render_checks: bool,
 }
 
+/// The strategy by which we will communicate with the Fluvio cluster
+#[derive(Debug, Clone)]
+pub enum IpMode {
+    NodePort,
+    ClusterIp,
+    LoadBalancer,
```
Comment (nicholastmosher) [inline src/cluster/src/start/k8.rs:369]: We may be able to get rid of LoadBalancer altogether

File: k8-util/helm/fluvio-app/values.yaml (hunk 1 above)
Comment (tjtelan) [reply k8-util/helm/fluvio-app/values.yaml:9]: The helm chart is optimized for local environments

File: Cargo.toml [hunk 4]
```diff
@@ -37,6 +37,8 @@ resolver = "2"
 [patch.crates-io]
 fluvio-future = { git = "https://github.com/infinyon/future-aio.git", branch = "master"}
 #fluvio-future = { path = "../future-aio" }
+#k8-types = { path = "../k8-api/src/k8-types" }
+k8-types = { git = "https://github.com/tjtelan/k8-api.git", branch = "add-node" }
```
Comment (sehz) [inline Cargo.toml:41]: should use release version?

File: src/cluster/src/start/k8.rs [hunk 5]
```diff
@@ -792,6 +799,57 @@ impl ClusterInstaller {
...
+            debug!("Using NodePort service type");
+            debug!("Getting external IP from K8s node");
+            let kube_client = &self.kube_client;
+
+            debug!("Trying to query for Nodes");
+
+            let nodes = kube_client
+                .retrieve_items::<NodeSpec, _>("")
+                .await
+                .expect("Node query failed");
```
Comment (sehz) [inline src/cluster/src/start/k8.rs:820]: remove `expect`

File: src/cluster/src/start/k8.rs [hunk 6]
```diff
@@ -792,6 +799,57 @@ impl ClusterInstaller {
...
+            for n in nodes.items.into_iter().map(|x| x.status.addresses) {
+                node_addr.extend(n)
+            }
+
+            debug!("Node Addresses: {:#?}", node_addr);
+
+            let external_addr = node_addr
+                .into_iter()
+                .find(|a| a.r#type == "InternalIP")
+                .expect("No nodes with InternalIP set");
```
Comment (sehz) [inline src/cluster/src/start/k8.rs:834]: remove `expect`

File: src/cluster/src/start/k8.rs [hunk 7]
```diff
@@ -792,6 +799,57 @@ impl ClusterInstaller {
...
+            ingress_address.insert("fluvio.io/ingress-address", external_addr.address);
+
+            let mut service_annotation = BTreeMap::new();
+            service_annotation.insert("serviceAnnotations", ingress_address);
+
+            let mut helm_lb_config = BTreeMap::new();
+            helm_lb_config.insert("loadBalancer", service_annotation);
+
+            serde_yaml::to_writer(&np_addr_fd, &helm_lb_config)
+                .expect("Failed to write ingress-addr to file");
```
Comment (sehz) [inline src/cluster/src/start/k8.rs:850]: remove `expect`

File: src/cluster/src/start/k8.rs [hunk 8]
```diff
@@ -931,23 +997,56 @@ impl ClusterInstaller {
...
                                     };
 
-                                    let ingress_addr = service
-                                        .status
-                                        .load_balancer
-                                        .ingress
-                                        .iter()
-                                        .find(|_| true)
-                                        .and_then(|ingress| ingress.host_or_ip().to_owned());
+                                    match service.spec.r#type.expect("Load Balancer Type") {
```
Comment (sehz) [inline src/cluster/src/start/k8.rs:1011]: remove expect

File: src/cluster/src/start/k8.rs [hunk 9]
```diff
@@ -931,23 +997,56 @@ impl ClusterInstaller {
...
+                                        LoadBalancerType::ClusterIP => {
+                                            return Ok(Some((format!("{}:{}",service.spec.cluster_ip,target_port),target_port)))
+                                        },
+                                        LoadBalancerType::NodePort => {
+                                            let node_port = node_port.expect("Expecting a NodePort port");
```
Comment (sehz) [inline src/cluster/src/start/k8.rs:1016]: remove expect

File: src/cluster/src/start/k8.rs [hunk 10]
```diff
@@ -931,23 +997,56 @@ impl ClusterInstaller {
...
+                                            let nodes = self.kube_client.retrieve_items::<NodeSpec, _>(ns).await?;
+                                            debug!("Output from k8 node query: {:#?}", &nodes);
+
+                                            let mut node_addr : Vec<NodeAddress> = Vec::new();
+                                            for n in nodes.items.into_iter().map(|x| x.status.addresses ) {
+                                                node_addr.extend(n)
+                                            }
 
+                                            // Return the first node with type "InternalIP"
+                                            let external_addr = node_addr.into_iter().find(|a| a.r#type == "InternalIP").expect("No nodes with InternalIP set");
```
Comment (sehz) [inline src/cluster/src/start/k8.rs:1028]: remove expect

File: src/sc/src/k8/controllers/spu.rs [hunk 11]
```diff
@@ -196,34 +196,61 @@ impl SpuController {
...
-                    ip: None,
-                });
-            }
-        }
 
-        if &computed_spu_ingress != spu_ingress {
+        // Choose the external port based on the service type
+        let mut computed_spu_ingressport = match lb_type {
+            Some(LoadBalancerType::NodePort) => IngressPort {
+                port: spu_svc_nodeport.expect("Expected a node port"),
```
Comment (sehz) [inline src/sc/src/k8/controllers/spu.rs:227]: remove expect

**Review by sehz — CHANGES_REQUESTED** [review]
Excellent.  Few nits.  Also make sure this works in the cloud env

**Comment by sehz** [comment]
Remove `tunnel` in the `k8-util/minikube/reset-minikube`

File: Cargo.toml (hunk 4 above)
Comment (tjtelan) [reply Cargo.toml:41]: Yeah, we should. I have a PR open and can make this change soon. https://github.com/infinyon/k8-api/pull/51

**Comment by sehz** [comment]
rebase

**Review by nacardin — APPROVED** [review]
Looks good, seems to respect `fluvio.io/ingress-address`.

**Review by sehz — APPROVED** [review]
Nice work.  please rebase it, so we can merge it

**Pull request closed without merging**